MSAL/src-internal/ios/MSALLegacySharedAccountsProvider.m

```diff
@@ -0,0 +1,512 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSALLegacySharedAccountsProvider.h"
+#import "MSIDKeychainTokenCache.h"
+#import "MSIDCacheKey.h"
+#import "MSIDCacheItemJsonSerializer.h"
+#import "MSALLegacySharedAccountFactory.h"
+#import "MSIDJsonObject.h"
+#import "MSALLegacySharedAccount.h"
+#import "MSALAccountEnumerationParameters.h"
+#import "MSIDConstants.h"
+#import "MSALErrorConverter.h"
+
+@interface MSALLegacySharedAccountsProvider()
+
+@property (nonatomic) MSIDKeychainTokenCache *keychainTokenCache;
+@property (nonatomic) NSString *serviceIdentifier;
+@property (nonatomic) NSString *applicationIdentifier;
+@property (nonatomic) dispatch_queue_t synchronizationQueue;
+
+@end
+
+@implementation MSALLegacySharedAccountsProvider
+
+#pragma mark - Init
+
+- (instancetype)initWithSharedKeychainAccessGroup:(NSString *)sharedGroup
+                                serviceIdentifier:(NSString *)serviceIdentifier
+                            applicationIdentifier:(NSString *)applicationIdentifier
+{
+    self = [super init];
+    
+    if (self)
+    {
+        self.keychainTokenCache = [[MSIDKeychainTokenCache alloc] initWithGroup:sharedGroup];
+        self.serviceIdentifier = serviceIdentifier;
+        self.applicationIdentifier = applicationIdentifier;
+        
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.legacysharedaccountsprovider-%@", [NSUUID UUID].UUIDString];
+        _synchronizationQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
```

Comment:
nit: Can we use self.synchronizationQueue here as well for consistency.

Reply:
sure